### PR TITLE
feat: project pipeline and pipeline definition

### PR DESCRIPTION
**Pull request by kakj-go**
#### What this PR does / why we need it:

-Move the application level pipeline to the project level pipeline;
-Show the pipeline in the form of instance list;
-pipeline definition and source idea

#### Which issue(s) this PR fixes:

- [Erda Cloud Issue Link](https://terminus-org.app.terminus.io/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJpdGVyYXRpb24iOls4ODMsNzcyXSwibWVtYmVyIjpbIjEwMDA1NjAiXX0%3D&id=265611&iterationID=772&pId=0&type=REQUIREMENT)

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |       Project pipeline interface modification  and add pipeline definition    |
| 🇨🇳 中文    |       项目流水线界面修改和增加流水线定义     |

### CODE REVIEW

File: apistructs/permission.go [hunk 1]
```diff
@@ -72,6 +72,7 @@ const (
 	AutotestSceneResource      string = "autotest-scene"
 	SceneSetResource           string = "sceneset"
 	CustomAddonResource        string = "customAddon"
+	ProjectPipelineResource    string = "projectPipeline"
```
Comment (Effet) [inline apistructs/permission.go:75]: afak, this new permission resource is not used, you should delete it

File: pkg/erda-configs/permission/backend/sys/org/project/project.yml [hunk 2]
```diff
@@ -839,4 +839,25 @@
...
+
+- role: Owner,Lead,PM,PD,Dev,QA,Ops
+  scope: project
+  resource: projectPipeline
+  action: UPDATE
+
+- role: Owner,Lead,PM,PD,Dev,QA,Ops
+  scope: project
+  resource: projectPipeline
+  action: DELETE
```
Comment (Effet) [inline pkg/erda-configs/permission/backend/sys/org/project/project.yml:863]: afak, this new permission resource is not used, you should delete it

File: conf/pipeline/pipeline.yaml [hunk 3]
```diff
@@ -16,6 +16,7 @@ mysql-xorm:
   username: "${MYSQL_USERNAME:root}"
   password: "${MYSQL_PASSWORD:123456}"
   database: "${MYSQL_DATABASE:test}"
+  show_sql: true
```
Comment (Effet) [inline conf/pipeline/pipeline.yaml:19]: I suggest to disable show_sql? will it cause performance problem?

File: conf/dop/dop.yaml [hunk 4]
```diff
@@ -39,7 +39,10 @@ erda.core.pipeline.cms-client:
 grpc-client@erda.core.pipeline.definition:
   addr: "${PIPELINE_GRPC_ADDR:pipeline:30810}"
 erda.core.pipeline.definition-client:
-erda.core.pipeline.definition-process-client:
```
Comment (Effet) [inline conf/dop/dop.yaml:42]: why just leave `definition-client` and remove `definition-process-client`? what difference are they?

File: conf/dop/i18n/cp/scenarios/project-pipeline-exec-list.yaml [hunk 5]
```diff
@@ -0,0 +1,17 @@
+zh:
+  pipelineName: 流水线名称
```
Comment (Effet) [inline conf/dop/i18n/cp/scenarios/project-pipeline-exec-list.yaml:2]: these changes is cp related, you should move to another pull request

File: .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql [hunk 6]
```diff
@@ -0,0 +1,54 @@
+CREATE TABLE `pipeline_definitions` (
```
Comment (Effet) [inline .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql:1]: as the standard defined, you should use `pipeline_definition` as table name

File: .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  `pipeline_source_id` varchar(36) NOT NULL DEFAULT '' COMMENT 'pipeline 来源id',
+  `pipeline_definition_extra_id` varchar(36) NOT NULL DEFAULT '' COMMENT '详情 id',
+  `category` varchar(20) NOT NULL DEFAULT '' COMMENT '类型',
+  `status` varchar(20) NOT NULL COMMENT '最后一次流水线执行状态',
+  `pipeline_id` bigint(20) DEFAULT NULL COMMENT '最后一次执行的流水线id',
+  PRIMARY KEY (`id`),
+  KEY `pipeline_source_id_index` (`pipeline_source_id`),
+  KEY `pipeline_id_index` (`pipeline_id`),
+  KEY `pipeline_definition_extra_id_index` (`pipeline_definition_extra_id`),
+  KEY `name_index` (`name`)
```
Comment (Effet) [inline .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql:21]: every index should place soft_deleted_at as first column, for example:
```sql
KEY `pipeline_source_id_index` (`soft_deleted_at`, `pipeline_source_id`),
```

File: .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COMMENT='pipeline definition table';
+
+
+CREATE TABLE `pipeline_definition_extras` (
```
Comment (Effet) [inline .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql:25]: `pipeline_definition_extra` as standard

File: .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  `id` varchar(36) NOT NULL DEFAULT '' COMMENT '主键',
+  `extra` mediumtext NOT NULL COMMENT '详细信息',
+  `created_at` datetime NOT NULL DEFAULT '1970-01-01 00:00:00' COMMENT '创建时间',
+  `updated_at` datetime NOT NULL DEFAULT '1970-01-01 00:00:00' ON UPDATE CURRENT_TIMESTAMP COMMENT '更新时间',
+  PRIMARY KEY (`id`)
```
Comment (Effet) [inline .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql:30]: also have a soft_deleted_at column

File: .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COMMENT='pipeline definition extra table';
+
+CREATE TABLE `pipeline_sources` (
```
Comment (Effet) [inline .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql:33]: `pipeline_source` as standard

File: .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  `pipeline_yml` mediumtext NOT NULL COMMENT 'yml 文件内容',
+  `remote` varchar(200) NOT NULL DEFAULT '' COMMENT '源地址',
+  `ref` varchar(200) NOT NULL DEFAULT '' COMMENT '源地址位置',
+  `path` varchar(200) NOT NULL DEFAULT '' COMMENT '源地址位置目录位置',
+  `name` varchar(200) NOT NULL DEFAULT '' COMMENT '流水线名称',
+  PRIMARY KEY (`id`),
+  KEY `source_type_index` (`source_type`),
+  KEY `remote_index` (`remote`),
+  KEY `name_index` (`name`),
+  KEY `ref_index` (`ref`)
```
Comment (Effet) [inline .erda/migrations/pipeline/20220112-pipeline-definition-and-source.sql:49]: missing soft_deleted_at in index

File: bundle/app.go [hunk 12]
```diff
@@ -47,7 +47,7 @@ func (b *Bundle) GetApp(id uint64) (*apistructs.ApplicationDTO, error) {
 	return &fetchResp.Data, nil
 }
 
-func (b *Bundle) GetMyApps(userid string, orgid uint64) (*apistructs.ApplicationListResponseData, error) {
+func (b *Bundle) GetMyApps(userid string, orgid, projectID uint64) (*apistructs.ApplicationListResponseData, error) {
```
Comment (Effet) [inline bundle/app.go:50]: please keep the old func, and add a new func, unless, orchestrator will be affected.

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/dop/providers/projectpipeline/provider.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package projectpipeline
+
+import (
+	"github.com/jinzhu/gorm"
+	"reflect"
```
Comment (Effet) [inline modules/dop/providers/projectpipeline/provider.go:19]: bad imports

**Review by Effet — CHANGES_REQUESTED** [review]

File: apistructs/pipeline.go [hunk 14]
```diff
@@ -342,8 +347,42 @@ type PipelinePageListRequest struct {
...
 	// internal use
 	SelectCols []string `schema:"-" ` // 需要赋值的字段列表，若不声明，则全赋值
-	AscCols    []string `schema:"-"`
-	DescCols   []string `schema:"-"`
+	AscCols    []string `schema:"ascCols"`
+	DescCols   []string `schema:"descCols"`
+
+	// pipeline definition search
+	PipelineDefinitionRequest     *PipelineDefinitionRequest
+	PipelineDefinitionRequestJSON string `schema:"pipelineDefinition"`
```
Comment (sfwn) [inline apistructs/pipeline.go:355]: PipelineDefinitionRequestJSONBase64

File: apistructs/pipeline_cron.go [hunk 15]
```diff
@@ -19,10 +19,11 @@ import (
...
 type PipelineCronPagingRequest struct {
-	AllSources bool             `schema:"allSources"`
-	Sources    []PipelineSource `schema:"source"`  // ?source=cdp-dev&source=cdp-test
-	YmlNames   []string         `schema:"ymlName"` // ?ymlName=11&ymlName=22
-	Enable     *bool            `schema:"enable"`
+	AllSources               bool             `schema:"allSources"`
+	Sources                  []PipelineSource `schema:"source"`  // ?source=cdp-dev&source=cdp-test
+	YmlNames                 []string         `schema:"ymlName"` // ?ymlName=11&ymlName=22
+	Enable                   *bool            `schema:"enable"`
+	PipelineDefinitionIDList []string         `schema:"pipelineDefinitionIDList"`
```
Comment (sfwn) [inline apistructs/pipeline_cron.go:26]: `schema: pipelineDefinitionID`

File: apistructs/pipeline_cron.go [hunk 16]
```diff
@@ -43,17 +44,19 @@ type PipelineCronDTO struct {
...
+	CronExpr               string         `json:"cronExpr"`
+	CronStartTime          *time.Time     `json:"cronStartTime"`
+	PipelineYmlName        string         `json:"pipelineYmlName"` // 一个分支下可以有多个 pipeline 文件，每个分支可以有单独的 cron 逻辑
+	BasePipelineID         uint64         `json:"basePipelineID"`  // 用于记录最开始创建出这条 cron 记录的 pipeline id
+	Enable                 *bool          `json:"enable"`          // 1 true, 0 false
+	PipelineYml            string         `json:"pipelineYml"`
+	ConfigManageNamespaces []string       `json:"configManageNamespaces"`
+	UserID                 string         `json:"userID"`
+	OrgID                  uint64         `json:"orgID"`
+	PipelineDefinitionID   string         `json:"pipelineDefinitionId"`
```
Comment (sfwn) [inline apistructs/pipeline_cron.go:58]: ID

Comment (sfwn) [reply apistructs/pipeline_cron.go:58]: search all Id and replace to ID

**Review by kakj-go — COMMENTED** [review]
review

File: conf/dop/dop.yaml (hunk 4 above)
Comment (kakj-go) [reply conf/dop/dop.yaml:42]: process was pipeline service client not grpc client, pipeline process service was removed

File: modules/pipeline/dbclient/op_pipeline.go [hunk 17]
```diff
@@ -352,45 +382,74 @@ func (client *Client) PageListPipelines(req apistructs.PipelinePageListRequest,
...
 	if len(req.TriggerModes) > 0 {
-		baseSQL.In("trigger_mode", req.TriggerModes)
+		baseSQL.In(tableFieldName((&spec.PipelineBase{}).TableName(), "trigger_mode"), req.TriggerModes)
 	}
 	if len(req.ClusterNames) > 0 {
-		baseSQL.In("cluster_name", req.ClusterNames)
+		baseSQL.In(tableFieldName((&spec.PipelineBase{}).TableName(), "cluster_name"), req.ClusterNames)
 	}
-	baseSQL.Where("is_snippet = ?", req.IncludeSnippet)
+	baseSQL.Where(tableFieldName((&spec.PipelineBase{}).TableName(), "is_snippet")+" = ?", req.IncludeSnippet)
```
Comment (sfwn) [inline modules/pipeline/dbclient/op_pipeline.go:424]: why there is a "+" somewhere?

File: modules/pipeline/initialize.go [hunk 18]
```diff
@@ -209,14 +207,14 @@ func (p *provider) do() error {
...
 
-	p.ReconcilerElection.OnLeader(func(ctx context.Context) {
-		engine.StartReconciler(ctx)
-		pipelineSvc.DoCrondAbout(ctx)
-	})
-
-	p.GcElection.OnLeader(func(ctx context.Context) {
-		engine.StartGC(ctx)
-	})
+	//p.ReconcilerElection.OnLeader(func(ctx context.Context) {
```
Comment (sfwn) [inline modules/pipeline/initialize.go:210]: ...

File: pkg/limit_sync_group/work.go [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+func (that *Worker) AddFunc(fun func(*Locker, ...interface{}) error, params ...interface{}) *Worker {
+	that.functions = append(that.functions, function{
+		function: fun,
+		params:   params,
+	})
+	return that
+}
+
+func (that *Worker) Do() *Worker {
```
Comment (sfwn) [inline pkg/limit_sync_group/work.go:55]: add unit-test for pkg.

File: pkg/time/mysql_time/mysql_time.go [hunk 20]
```diff
@@ -17,6 +17,6 @@ package mysql_time
 import "time"
 
 func GetMysqlDefaultTime() *time.Time {
-	date := time.Date(1000, 01, 01, 0, 0, 0, 0, time.Local)
+	date := time.Date(1970, 01, 01, 0, 0, 0, 0, time.UTC)
```
Comment (sfwn) [inline pkg/time/mysql_time/mysql_time.go:20]: any other invoker will be influenced?

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/dop/providers/projectpipeline/deftype/batch_run.go [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deftype
+
+import "github.com/erda-project/erda/apistructs"
+
+type ProjectPipelineBatchRun struct {
+	PipelineDefinitionIDS []string
```
Comment (Effet) [inline modules/dop/providers/projectpipeline/deftype/batch_run.go:20]: PipelineDefinitionIDs

File: modules/pipeline/spec/pipeline_base.go [hunk 22]
```diff
@@ -52,8 +54,20 @@ type PipelineBase struct {
 
 	TimeCreated *time.Time `json:"timeCreated,omitempty" xorm:"created"`
 	TimeUpdated *time.Time `json:"timeUpdated,omitempty" xorm:"updated"`
+
+	PipelineDefinitionId string `json:"pipelineDefinitionId"`
```
Comment (Effet) [inline modules/pipeline/spec/pipeline_base.go:58]: PipelineDefinitionID

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/pipeline/dbclient/op_pipeline.go (hunk 17 above)
Comment (kakj-go) [reply modules/pipeline/dbclient/op_pipeline.go:424]: Join query is added, and all corresponding query fields must be spliced with table names

File: pkg/time/mysql_time/mysql_time.go (hunk 20 above)
Comment (kakj-go) [reply pkg/time/mysql_time/mysql_time.go:20]: Because the database specification requires a default time, the time zone of the previous default time is incorrect

File: api/proto/core/pipeline/definition/definition.proto [hunk 23]
```diff
@@ -91,7 +92,6 @@ message PipelineDefinitionUpdateRequest {
   string executor = 3;
   string pipelineSourceId = 4;
   string category = 5;
-  PipelineDefinitionExtra extra = 6;
   google.protobuf.Timestamp startedAt = 7;
```
Comment (sfwn) [inline api/proto/core/pipeline/definition/definition.proto:95]: adjust order

File: apistructs/pipeline.go [hunk 24]
```diff
@@ -342,8 +347,42 @@ type PipelinePageListRequest struct {
...
 	// internal use
 	SelectCols []string `schema:"-" ` // 需要赋值的字段列表，若不声明，则全赋值
-	AscCols    []string `schema:"-"`
-	DescCols   []string `schema:"-"`
+	AscCols    []string `schema:"ascCol"`
+	DescCols   []string `schema:"descCol"`
+
+	// pipeline definition search
+	PipelineDefinitionRequest           *PipelineDefinitionRequest
+	PipelineDefinitionRequestJSONBase64 string `schema:"pipelineDefinition"`
```
Comment (sfwn) [inline apistructs/pipeline.go:355]: adjust schema key

**Review by sfwn — APPROVED** [review]

File: api/proto/core/pipeline/definition/definition.proto (hunk 23 above)
Comment (kakj-go) [reply api/proto/core/pipeline/definition/definition.proto:95]: After this PR merge, I will propose to change the PR, because the detection time is too long

File: apistructs/pipeline.go (hunk 24 above)
Comment (kakj-go) [reply apistructs/pipeline.go:355]: For the time being, there is no good scheme for receiving values. All fields of definition can be moved out. If necessary, they can be done later. Let's shelve it first

File: api/proto/core/pipeline/definition/definition.proto (hunk 23 above)
Comment (kakj-go) [reply api/proto/core/pipeline/definition/definition.proto:95]: https://github.com/erda-project/erda/pull/3712/files
